File: Excel/1000-1099/1032/Challenge 1032.R
library(tidyverse)
library(readxl)

path <- "1000-1099/1032/1032 Marks Calculation.xlsx"
marks <- read_excel(path, range = "A2:D30")
expected <- read_excel(path, range = "F2:G7")

score <- function(d) {
  if (nrow(d) < 5) {
    return(0)
  }
  core <- d %>% filter(Subject_Type == "Core") %>% pull(Marks)
  electives <- d %>%
    filter(Subject_Type == "Elective") %>%
    arrange(desc(Marks)) %>%
    slice_head(n = 5 - length(core)) %>%
    pull(Marks)
  failing <- sort(core[core < 50])
  n <- min(length(failing), length(electives))
  electives[seq_len(n)] <- electives[seq_len(n)] * .9
  round(sum(core) + sum(electives))
}

result <- marks %>%
  group_by(Student_ID) %>%
  summarise(Final_Score = score(pick(everything())), .groups = "drop")

print(all.equal(result, expected))

File: Excel/1000-1099/1032/Challenge 1032.py
import pandas as pd

path = "1000-1099/1032/1032 Marks Calculation.xlsx"
marks = pd.read_excel(path, usecols="A:D", skiprows=1, nrows=28)
expected = pd.read_excel(path, usecols="F:G", skiprows=1, nrows=5)
expected.columns = ["Student_ID", "Final_Score"]


def score(d):
    if len(d) < 5:
        return 0
    core = d.loc[d.Subject_Type.eq("Core"), "Marks"].tolist()
    electives = sorted(d.loc[d.Subject_Type.eq("Elective"), "Marks"], reverse=True)[
        : 5 - len(core)
    ]
    failing = sorted(mark for mark in core if mark < 50)
    for i in range(min(len(failing), len(electives))):
        electives[i] *= 0.9
    return round(sum(core) + sum(electives))


result = (
    marks.groupby("Student_ID", sort=True)
    .apply(score, include_groups=False)
    .rename("Final_Score")
    .reset_index()
)
print(result.equals(expected))
